File: app/src/main/java/knf/kuma/database/dao/DownloadsDAO.java
package knf.kuma.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.paging.DataSource;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

import knf.kuma.pojos.DownloadObject;

/**
 * Created by Jordy on 10/01/2018.
 */

@Dao
public interface DownloadsDAO {
    @Query("SELECT * FROM downloadobject")
    DataSource.Factory<Integer,DownloadObject> getAll();

    @Query("SELECT * FROM downloadobject WHERE eid LIKE :eid")
    DownloadObject getByEid(String eid);

    @Query("SELECT * FROM downloadobject WHERE file LIKE :name")
    DownloadObject getByFile(String name);

    @Query("SELECT * FROM downloadobject WHERE eid LIKE :eid")
    LiveData<DownloadObject> getLiveByEid(String eid);

    @Query("SELECT * FROM downloadobject WHERE `key` LIKE :key")
    LiveData<DownloadObject> getLiveByKey(int key);

    @Query("SELECT * FROM downloadobject WHERE state<=0")
    LiveData<List<DownloadObject>> getActive();

    @Query("SELECT count(*) FROM downloadobject WHERE state=-1")
    int countPending();

    @Query("DELETE FROM downloadobject WHERE eid LIKE :eid")
    void deleteByEid(String eid);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(DownloadObject object);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void update(DownloadObject object);

    @Delete
    void delete(DownloadObject object);
}

File: app/src/main/java/knf/kuma/database/dao/DownloadsDAO.kt
package knf.kuma.database.dao

import androidx.lifecycle.LiveData
import androidx.paging.DataSource
import androidx.room.*
import knf.kuma.pojos.DownloadObject

@Dao
interface DownloadsDAO {
    @get:Query("SELECT * FROM downloadobject")
    val all: DataSource.Factory<Int, DownloadObject>

    @get:Query("SELECT * FROM downloadobject WHERE state<=0")
    val active: LiveData<List<DownloadObject>>

    @Query("SELECT * FROM downloadobject WHERE eid LIKE :eid")
    fun getByEid(eid: String): DownloadObject?

    @Query("SELECT * FROM downloadobject WHERE did = :did")
    fun getByDid(did: Int): DownloadObject?

    @Query("SELECT * FROM downloadobject WHERE file LIKE :name")
    fun getByFile(name: String): DownloadObject?

    @Query("SELECT * FROM downloadobject WHERE eid = :eid")
    fun getLiveByEid(eid: String): LiveData<DownloadObject>

    @Query("SELECT * FROM downloadobject WHERE `key` LIKE :key")
    fun getLiveByKey(key: Int): LiveData<DownloadObject>

    @Query("SELECT count(*) FROM downloadobject WHERE state=-1")
    fun countPending(): Int

    @Query("SELECT count(*) FROM downloadobject WHERE state=-1 OR state=0")
    fun countActive(): Int

    @Query("DELETE FROM downloadobject WHERE eid LIKE :eid")
    fun deleteByEid(eid: String)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(downloadObject: DownloadObject)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    fun update(downloadObject: DownloadObject)

    @Delete
    fun delete(downloadObject: DownloadObject)
}
